iree/compiler/Translation/SPIRV/LinalgToSPIRV/SplitDispatchFunctionPass.cpp

```diff
@@ -0,0 +1,239 @@
+  SmallVector<linalg::LinalgOp, 4> linalgOps =
+      llvm::to_vector<4>(oldFn.getBlocks().front().getOps<linalg::LinalgOp>());
+  if (!canDistributeLinalgOps(linalgOps)) {
+    return rewriter.notifyMatchFailure(
+        oldFn, "cannot distribute Linalg ops into multiple kernels");
+  }
+
+  ModuleOp moduleOp = cast<ModuleOp>(oldFn.getParentOp());
+  Block &oldFnBlock = oldFn.getBlocks().front();
+  Location loc = oldFn.getLoc();
+
+  SmallVector<std::string, 4> splitKernels;
+  splitKernels.reserve(linalgOps.size());
+  llvm::SmallPtrSet<Operation *, 16> closure;
+
+  for (const auto &linalgOp : llvm::enumerate(linalgOps)) {
+    // Createa a new function for hosting this Linalg op.
+    splitKernels.emplace_back(
+        llvm::formatv("{0}_dispatch_{1}", oldFn.getName(), linalgOp.index()));
+    StringRef newFnName = splitKernels.back();
+    rewriter.setInsertionPointToStart(moduleOp.getBody());
+    auto newFn = rewriter.create<FuncOp>(loc, newFnName, oldFn.getType(),
+                                         /*attrs=*/ArrayRef<NamedAttribute>());
+
+    // Copy over all attributes except type and name.
+    for (const auto &namedAttr : oldFn.getAttrs()) {
+      if (namedAttr.first != impl::getTypeAttrName() &&
+          namedAttr.first != SymbolTable::getSymbolAttrName())
+        newFn.setAttr(namedAttr.first, namedAttr.second);
+    }
+
+    // Collect the closure for the current Linalg op.
+    closure.clear();
+    collectClosure(linalgOp.value(), closure);
+
+    // Clone all ops in the closure to the new function.
+    Block *newFnBlock = newFn.addEntryBlock();
+    rewriter.setInsertionPointToStart(newFnBlock);
+    BlockAndValueMapping remapper;
+    for (Operation &op : oldFnBlock) {
+      if (closure.count(&op) == 0) continue;
+      rewriter.insert(op.clone(remapper));
+      if (&op == linalgOp.value()) break;
+    }
+    rewriter.insert(oldFnBlock.getTerminator()->clone(remapper));
+  }
+
+  // Add the entry point schedule to the module op.
+  SmallVector<Attribute, 4> entryPoints;
+  entryPoints.reserve(linalgOps.size());
+  for (const std::string &kernel : splitKernels) {
+    entryPoints.emplace_back(rewriter.getStringAttr(kernel));
+  }
+  moduleOp.setAttr(getEntryPointScheduleAttrName(),
+                   rewriter.getArrayAttr(entryPoints));
+
+  rewriter.eraseOp(oldFn);
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Pass entry point and registration
+//===----------------------------------------------------------------------===//
+
+std::unique_ptr<OperationPass<ModuleOp>> createSplitDispatchFunctionPass() {
+  return std::make_unique<SplitDispatchFunctionPass>();
+}
+
+static PassRegistration<SplitDispatchFunctionPass> pass(
+    "iree-codegen-split-dispatch-function",
+    "Split workload to multiple dispatch functions to satisfy computation "
+    "dependency for GPU lowering");
+}  // namespace iree_compiler
```

Comment:
vertical whitespace between namespace names helps readability (here and at the top)